pkg/workflows/wasm/host/module.go

```diff
@@ -580,7 +603,73 @@ func runWasm[I idMessage, O proto.Message](
 }
 
 func (m *module) SetCapabilityExecutor(handler CapabilityExecutor) error {
-	return errors.New("not implemented")
+	m.handlerLock.Lock()
+	defer m.handlerLock.Unlock()
+
+	if handler == nil {
+		return fmt.Errorf("invalid handler (%T): cannot set nil call capability handler", handler)
+	}
+
+	m.handler = handler
+
+	return nil
+}
+
+// callCapAsync async calls a capability by placing execution results onto a
+// channel and storing each channel with a unique identifier for future
+// retrieval on await.
+func (m *module) callCapAsync(ctx context.Context, req *sdkpb.CapabilityRequest) ([sdk.IdLen]byte, error) {
+	callId := uuid.NewString()
+
+	go func(id string) {
+		ch := make(chan *sdkpb.CapabilityResponse, 1)
+		m.capabilityCallStore.add(id, ch)
+
+		m.handlerLock.RLock()
+		resp, err := m.handler.CallCapability(ctx, req)
+		m.handlerLock.RUnlock()
+
+		if err != nil {
+			resp = &sdkpb.CapabilityResponse{
+				Response: &sdkpb.CapabilityResponse_Error{
+					Error: err.Error(),
+				},
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+		case ch <- resp:
+		}
+	}(callId)
+
+	var idBuffer [sdk.IdLen]byte
+	copy(idBuffer[:], []byte(callId))
+	return idBuffer, nil
+}
+
+func (m *module) awaitCapabilities(ctx context.Context, acr *sdkpb.AwaitCapabilitiesRequest) (*sdkpb.AwaitCapabilitiesResponse, error) {
+	responses := make(map[string]*sdkpb.CapabilityResponse, len(acr.Ids))
+
+	for _, callId := range acr.Ids {
+		ch, err := m.capabilityCallStore.get(callId)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get call from store : %w", err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil, fmt.Errorf("failed to wait for capability response %s : %w", callId, ctx.Err())
```

Comment:
on its own this message will not indicate which capability was timed out, maybe add the request to the item put in the capabilityCallStore so it can be add to this message (the engine could arguably do this as well, but here would be eaiser I think)